math/mathcore/inc/Math/Functor.h

```diff
@@ -163,6 +163,16 @@ class GradFunctor : public IGradientFunctionMultiDim  {
       : fDim{dim}, fFunc{std::bind(memFn, p, std::placeholders::_1)}, fDerivFunc{std::bind(gradFn, p, std::placeholders::_1, std::placeholders::_2)}
    {}
 
+   /// Construct from a pointer to member function and member function, types for function and full derivative
+   /// evaluations.
+   template <class PtrObj, typename MemFn, typename FullGradMemFn>
+   GradFunctor(unsigned int dim, const PtrObj &p, MemFn memFn, FullGradMemFn gradFn)
```

Comment:
The other constructor with the member functions takes the `dim` parameter as the last one. It would be good to be more consistent also do this here. Also, I think it makes more sense to not call the type `FullGradMemFn`, but just `GradMemFn`, and to avoid confusion better rename the type in the other constructor to `DerivMemFn`. Like this, we have one constructor with `DerivMemFn` and one with `GradMemFn` that are reasonably self explanatory.

Reply:
We have to either rearrange the params or specialize the template otherwise it will be an ambiguous call between this and the the other ctor that takes the partial derivative function.

The renaming suggestions make sense though

Reply:
It's true that keeping a nice and consistent interface here is a bit more work, but I think with these fundamental `ROOT::Math` interfaces we should not slack. The `GradMemFn` and `DerivMemFn` will have different return types. I think based on this you can resolve the ambiguity with [`std::enable_if`](https://en.cppreference.com/w/cpp/types/enable_if) (in fact it is exactly meant for this situation, maybe you didn't know about this nice feature yet :D)

Reply:
I do know about enable_if just wasn't sure if the interface functions were consistent enough to use that.